### PR TITLE
Addressing-7747;-_.merge-=>-Object.assign

### PR DESCRIPTION
**Pull request by pjmattingly**
Addresses: #7747

> Replace \_.merge with Object.assign. Note they're not equivalent (_.merge does deep merge), so it should be done with care in all places where it's safe to rely on Object.assign instead (majority of cases)

### CODE REVIEW

**Comment by pjmattingly** [comment]
some _.merge instance can't be replaced with Object.assgn; Those files are:

```
/lib/classes/service.js
    /lib/plugins/aws/package/compile/events/cloud-front.js
    /lib/plugins/aws/package/compile/events/cognito-user-pool.js
    /lib/plugins/aws/package/compile/events/alb/lib/permissions.js
    /lib/plugins/aws/package/compile/events/api-gateway/lib/usage-plan.js
    /lib/plugins/aws/package/compile/events/api-gateway/lib/method/index.js
    /lib/plugins/aws/package/lib/merge-custom-provider-resources.js
```

Some files had no tests, and so were not modifed; Those files are:

```
/lib/plugins/aws/package/compile/events/api-gateway/lib/method/responses.js
    /lib/plugins/aws/package/compile/events/websockets/lib/api.js
    /test/utils/api-gateway.js
```

Some test files use _.merge, so those were skipped as they weren't in customer-facing code:

```
/test/unit/lib/configuration/variables/sources/instance-dependent/get-aws.test.js
    /test/unit/lib/configuration/variables/sources/instance-dependent/get-sls.test.js
```

File: lib/classes/config.js [hunk 1]
```diff
@@ -12,7 +11,7 @@ class Config {
   }
 
   update(config) {
-    return _.merge(this, config);
+    return Object.assign(this, config);
```
Comment (pgrzesik) [inline lib/classes/config.js:14]: Did you validate that it's going to be a safe change? It seems like a pretty generic method where deep vs shallow copy behavior might have an impact

File: lib/classes/service.js [hunk 2]
```diff
@@ -284,7 +284,7 @@ class Service {
   }
 
   update(data) {
-    return _.merge(this, data);
+    return Object.assign(this, data);
```
Comment (pgrzesik) [inline lib/classes/service.js:287]: Is it safe to use `Object.assign` here? Is the deep merge not needed here?

File: lib/classes/service.js [hunk 3]
```diff
@@ -346,7 +346,7 @@ class Service {
 
   publish(dataParam) {
     const data = dataParam || {};
-    this.pluginsData = _.merge(this.pluginsData, data);
+    this.pluginsData = Object.assign(this.pluginsData, data);
```
Comment (pgrzesik) [inline lib/classes/service.js:349]: Is it safe to use `Object.assign` here? Is the deep merge not needed here?

File: lib/plugins/aws/package/compile/events/cloud-front.js [hunk 4]
```diff
@@ -612,6 +612,7 @@ class AwsCompileCloudFrontEvents {
 
       Object.assign(Resources, { [this.cloudFrontDistributionLogicalId]: CloudFrontDistribution });
 
+      // #7747, could not replace _.merge with Object.assign, as this does deep merges
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/events/cloud-front.js:615]: I think these assignments look pretty safe to be refactored - do you think we should keep merge here?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thanks a lot @pjmattingly - it looks good in general! I have a few comments, please let me know what do you think 🙇

**Review by medikoo — COMMENTED** [review]
@pgrzesik @pjmattingly with this change (while it's highly desirable) we need to be very careful. 

If we touch not well tested areas it's possible we may introduce regressions, as behavior of `_.merge` and `Object.assign` differs.

Ideally, if each change is validated individually by looking through carefully what it does and what further logic it affects.

I would also add a comment in each case where we leave `_.merge` usage with reasoning, on why `_.merge` is wanted approach

@pjmattingly please confirm also by running integration tests, that we did not break anything

**Comment by pjmattingly** [comment]
@medikoo 
> please confirm also by running integration tests, that we did not break anything

Will do.

> I would also add a comment in each case where we leave _.merge usage with reasoning, on why _.merge is wanted approach

I'll find my notes and add some more detail in comments for the unchanged ones.

> Ideally, if each change is validated individually by looking through carefully what it does and what further logic it affects.

I was working on #11273 and I found a need for inspecting objects during testing (See also: https://www.reddit.com/r/javascript/comments/w7ial4/askjs_a_library_for_tracking_the_properties_of_an/?). I think such functionality might also be helpful here, as the properties for the object can be tracked during (integration) testing. Would you agree?

If so, then it might be worth making / finding code to track and report on the different values that effected objects go through and potentially where the code is called from. This way we can have more confidence about: 1) What properties and values the objects typically take and 2) What external code these changes touch.

What do you think?

**Comment by medikoo** [comment]
> If so, then it might be worth making / finding code to track and report on the different values that effected objects go through and potentially where the code is called from. 

I'm not sure if it's necessary to bundle such a thing into a test suite. It feels a bit overkill (and definitely adds to technical debt), but if in this specific case, it would help you test whether change is not breaking. Feel free to  set up such thing temporarily (just to test proposed changes, but not be a part of a PR)

**Comment by pjmattingly** [comment]
Exactly. Thanks for weighing in!

**Comment by pjmattingly** [comment]
Integration tests look good.

Next, adding notes about why leaving in some `_.merge()`s. Then, getting a closer look at each change to determine the inputs and outputs, and what other code is effected by these changes indirectly.

> I would also add a comment in each case where we leave _.merge usage with reasoning, on why _.merge is wanted approach

> Ideally, if each change is validated individually by looking through carefully what it does and what further logic it affects.

**Comment by medikoo** [comment]
@pjmattingly once it's ready for re-review, it's best if you simply re-request review in right top box

**Comment by pjmattingly** [comment]
Okay, will do. I as making notes to show progress toward completion; It helps me remember what I was doing with this PR.

Is that okay?

**Comment by medikoo** [comment]
> Okay, will do. I as making notes to show progress toward completion; It helps me remember what I was doing with this PR.
> Is that okay?

Yes, that's totally good. Great thanks for helping on that one 👍

**Pull request closed without merging**